File: php/src/Advertisement/Application/Command/UpdateAdvertisement/UpdateAdvertisementUseCase.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisement\Application\Command\UpdateAdvertisement;

use Demo\App\Advertisement\Application\Exceptions\InvalidPasswordException;
use Demo\App\Advertisement\Domain\AdvertisementRepository;
use Demo\App\Advertisement\Domain\Exceptions\AdvertisementNotFoundException;
use Demo\App\Advertisement\Domain\Model\Advertisement;
use Demo\App\Advertisement\Domain\Model\ValueObject\AdvertisementId;
use Demo\App\Advertisement\Domain\Model\ValueObject\Password;
use Chemem\Bingo\Functional\Functors\Monads\Either;
use Demo\App\Advertisement\Infrastructure\Exceptions\ZeroRecordsException;

final class UpdateAdvertisementUseCase
{
    public function __construct(private AdvertisementRepository $advertisementRepository)
    {
    }

    public function execute(UpdateAdvertisementCommand $command): Either
    {
        return AdvertisementId::build($command->id)
            ->flatMap(function($advertisementId) use ($command) {
                $advertisementResult = $this->advertisementRepository->findById($advertisementId);
                if ($advertisementResult->isLeft() && $advertisementResult->getLeft() instanceof ZeroRecordsException) {
                    return Either::left(AdvertisementNotFoundException::withId($advertisementId->value()));
                }
                return $advertisementResult
                    ->flatMap(function($advertisement) use ($command) {
                        if (!$advertisement->password()->isValidatedWith($command->password)){
                            return Either::left(InvalidPasswordException::build());
                        }
                        return Password::fromPlainPassword($command->password)
                            ->map(function($newPassword) use ($advertisement, $command) {
                                $advertisement->update(
                                    $command->description,
                                    $command->email,
                                    $newPassword,
                                );
                                $this->advertisementRepository->save($advertisement);
                                return Either::right(null);
                            });
                    });
            });
    }

    public function execute2(UpdateAdvertisementCommand $command): Either
    {
        $advertisementIdResult = AdvertisementId::build($command->id);

        if ($advertisementIdResult->isLeft()) {
            return $advertisementIdResult;
        }

        $advertisementId = $advertisementIdResult->getRight();
        $advertisementResult = $this->advertisementRepository->findById($advertisementId);

        if ($advertisementResult->isLeft() && $advertisementResult->getLeft() instanceof ZeroRecordsException) {
            return Either::left(AdvertisementNotFoundException::withId($advertisementId->value()));
        }

        $advertisement = $advertisementResult->getRight();

        if (!$advertisement->password()->isValidatedWith($command->password)){
            return Either::left(InvalidPasswordException::build());
        }

        $newPasswordResult = Password::fromPlainPassword($command->password);

        if ($newPasswordResult->isLeft()) {
            return $newPasswordResult;
        }

        $newPassword = $newPasswordResult->getRight();
        $advertisement->update($command->description, $command->email, $newPassword);
        $this->advertisementRepository->save($advertisement);

        return Either::right(null);
    }
}

File: kotlin/src/main/kotlin/advertisement/application/updateAdvertisement/UpdateAdvertisementUseCase.kt
package advertisement.application.updateAdvertisement

import advertisement.application.errors.PasswordDoesNotMatchError
import advertisement.domain.AdvertisementRepository
import advertisement.domain.errors.AdvertisementNotFoundError
import advertisement.domain.model.Advertisement
import advertisement.domain.model.value_object.AdvertisementId
import advertisement.domain.model.value_object.Description
import advertisement.domain.model.value_object.Password
import arrow.core.Either
import arrow.core.raise.either
import common.BoundedContextError

class UpdateAdvertisementUseCase(private val advertisementRepository: AdvertisementRepository) {
    fun execute(updateAdvertisementCommand: UpdateAdvertisementCommand): Either<BoundedContextError, Unit> {
        return either {
            val advertisementId = AdvertisementId.build(updateAdvertisementCommand.id).bind()
            val description = Description.build(updateAdvertisementCommand.description).bind()
            val advertisement = getAdvertisement(advertisementId).bind()

            validatePassword(advertisement, updateAdvertisementCommand.password).bind()

            val password = Password.fromPlainPassword(updateAdvertisementCommand.password).bind()
            val updatedAdvertisement = advertisement.update(description, password).bind()

            advertisementRepository.save(updatedAdvertisement).bind()
        }
    }

    private fun getAdvertisement(advertisementId: AdvertisementId): Either<BoundedContextError, Advertisement> {
        val advertisement = advertisementRepository.findById(advertisementId)
        if (advertisement.isLeft()) {
            return Either.Left(AdvertisementNotFoundError.withId(advertisementId.value()))
        }

        return Either.Right(advertisement.getOrNull()!!)
    }

    private fun validatePassword(advertisement: Advertisement, password: String): Either<BoundedContextError, Unit> {
        if (advertisement.password.isValidatedWith(password).not())
            return Either.Left(PasswordDoesNotMatchError.build())

        return Either.Right(Unit)
    }
}
